### DISCUSSION TITLE
Rails engine support

### DISCUSSION THREAD

**Discussion by flavorjones** [rails/tailwindcss-rails#355] [Ideas]
I wanted to start a discussion thread about support for Rails Engines and multiple Tailwind CSS files. This is an attempt to bring a few related conversations together and start to drive towards a clear understanding of the problem(s) people need solved and talk about potential solutions.

## Context

There have been a few issues and pull requests around a similar set of problems. In chronological order:

- https://github.com/rails/tailwindcss-rails/issues/108
- https://github.com/rails/tailwindcss-rails/issues/150
- https://stackoverflow.com/questions/71232601/how-to-use-tailwind-css-gem-in-a-rails-7-engine/
- https://github.com/rails/tailwindcss-rails/issues/151
- https://github.com/rails/tailwindcss-rails/pull/242

The topics from #108, #151, and #242 are related: how to support multiple CSS files, whether those are in engines/gems or elsewhere in the Rails application.

Issue #150 is slightly different and relates to full support for Rails engine generators. If we want to implement build-time support for Engines using tailwindcss, then we logically should implement generator support for Engines, too.

(The Stack Overflow thread touches on both topics.)

## The problem space

Here's my first draft of articulating a problem statement (feedback welcome):

> `tailwindcss-rails` should make it easy to write an Engine that uses Tailwind CSS.

The current tailwindcss-rails generator and its configuration both assume it's running in a top-level Rails application, and hardcodes some of those assumptions into, e.g., `config/tailwind.config.js`. The gem doesn't provide any ways to dynamically configure or extend this behavior.

As a result, people have fallen back on hacks like [generating `tailwind.config.js` dynamically](https://github.com/rails/tailwindcss-rails/issues/108#issuecomment-1017958397) or [forking out to bundler within `tailwind.config.js` to inject additional paths](https://stackoverflow.com/questions/71232601/how-to-use-tailwind-css-gem-in-a-rails-7-engine/74737193#74737193).

This problem statement doesn't explicitly include the "multiple Tailwind CSS files in an app" scenario from #242. In that discussion, @dhh gently pushed back on having multiple files in an app, which I think is healthy. But I suspect, if tailwindcss-rails provides a mechanism for building against assets in engines or gems, whatever mechanism we implement can be used by people to declare their separate Tailwind CSS files if they _really_ want to do it.

## The solution space

I think the broad strokes of how an engine integrates into railties are pretty well-known at this point, but broadly:

- the engine declares a dependency on the `tailwindcss-rails` gem
- the engine has its own tailwind config file
- which it registers in a Rails initializer, perhaps appending to `config.tailwindcss_rails.config_files`

Then, the gem's rake tasks `tailwindcss:build` and `tailwindcss:watch`, as well as the Puma plugin and/or the engine server hook from #347, would be extended to build multiple Tailwind CSS files.

We would also implement support for the plugin generator so that if an engine is created via `rails plugin new --css tailwind` then the above configuration is already in place.

## Next steps

Heading into Rails Conf in about 10 days, I'm going to be at the Hack Day and would like to work on this there.

I'd love feedback from y'all on either the problem statement (did I miss a use case? does the problem statement presume too much about the shape of the solution? are there alternatives you'd like to offer) or the solution space. Thank you!

**Comment by ianterrell** [comment]
> The topics from https://github.com/rails/tailwindcss-rails/issues/108, https://github.com/rails/tailwindcss-rails/issues/151, and https://github.com/rails/tailwindcss-rails/pull/242 are related: how to support multiple CSS files, whether those are in engines/gems or elsewhere in the Rails application. 
> ...
> I'd love feedback from y'all on either the problem statement 

I only want to chime in that in #108 I am not necessarily talking about building multiple CSS files per se, but rather allowing a single Tailwind generator to see all of an engine's views, partials, view components, and other files that contain Tailwind classes. When Tailwind generates its CSS file it scans all the files in the [content configuration](https://tailwindcss.com/docs/content-configuration) to know which utility classes it needs to include. 

I know everyone working on this knows the above — I'm just including it for thoroughness since I didn't read this issue as exactly speaking to that.

> the gem's rake tasks tailwindcss:build and tailwindcss:watch, as well as the Puma plugin and/or the engine server hook from https://github.com/rails/tailwindcss-rails/pull/347, would be extended to build multiple Tailwind CSS files.

I may be misinterpreting this, but — I think what you're saying is that each engine would know about its own files with its own config file, and Tailwind would build a CSS file per engine which would be accessible. I don't know that it would necessarily matter, but I find it preferable to have a single Tailwind-generated CSS file. I think we're the odd ducks, but my work's main app is built with something like a [Component-Based Rails Application](https://cbra.info) approach, and we have ~30 engines. There would end up being a lot of duplication in each Tailwind file for our use case, and it would be cumbersome in the layouts to figure out which Tailwind CSS files to include in the header when we render view components from other components (via dependency injection).

Hopefully I haven't misinterpreted anything too much — thank you & all for the work on this gem & topic!

**Reply by flavorjones** [reply]
@ianterrell Thanks for clarifying your needs! Hearing that you've got 30+ engines _and_ would want to render views from other engines is good information.

Maybe it would be better for engines to register their asset and view paths in an initializer, and then have just one tailwind css file? :thinking: I'm curious if this is how other folks feel, too!

**Reply by javierjulio** [reply]
@flavorjones hello and thank you. 👋🏻 As maintainer of ActiveAdmin, I'm interested in seeing this capability since with ActiveAdmin v4 beta it's now built with Tailwind but I wasn't able to use this gem successfully with an engine. [Some of our users have managed to hack in tailwindcss-rails support](https://github.com/activeadmin/activeadmin/discussions/8223). It would help to have the option to support multiple Tailwind config files so that way users with a large admin (or any large Rails engine) can have a separate Tailwind setup (e.g. different plugins) from their Rails app if desired.

**Reply by flavorjones** [reply]
@javierjulio Thank you for articulating your needs! The use case of a large admin is exactly what I had in mind when I made my original suggestion (multiple configs and multiple output CSS files. I really want to support this use case.

**Comment by dhh** [comment]
Would be happy to see engines supported, but imo the focus should be on expanding the files TW uses to look for classes in. Rather than trying to have multi outputs. I think in almost all cases, we're better off just producing one output for all classes found. And it'll be much easier, and we should be able to do so virtually without specific configuration.

**Reply by flavorjones** [reply]
OK, I'll spike on this approach.

**Reply by dhh** [reply]
Not saying we couldn't potentially also explore multiple outputs, but I'd want to see a clear case of where that makes a big difference before going down that path. And I think we get 90% of the value by simply allowing TW to produce the right output for all classes found in all dependencies.

**Reply by nickpoorman** [reply]
I think I can get onboard with this approach. I think I would need engine styles to be prefixed or namespaced under the engine somehow so they don't conflict.

**Reply by flavorjones** [reply]
OK, so I spiked on this approach and got something working that will scan engine files upon invocation of `rails tailwindcss:build`, then write all the tailwind CSS classes to `app/assets/builds/tailwind.css`.

The changes to this gem would look like:

```patch
diff --git a/lib/install/tailwind.config.js b/lib/install/tailwind.config.js
index d6ad82c0..16bbedda 100644
--- a/lib/install/tailwind.config.js
+++ b/lib/install/tailwind.config.js
@@ -1,12 +1,7 @@
 const defaultTheme = require('tailwindcss/defaultTheme')
 
 module.exports = {
-  content: [
-    './public/*.html',
-    './app/helpers/**/*.rb',
-    './app/javascript/**/*.js',
-    './app/views/**/*.{erb,haml,html,slim}'
-  ],
+  content: <%= Tailwindcss::Engine.globs.to_json %>,
   theme: {
     extend: {
       fontFamily: {
diff --git a/lib/tailwindcss/commands.rb b/lib/tailwindcss/commands.rb
index 08b70d33..9a66987e 100644
--- a/lib/tailwindcss/commands.rb
+++ b/lib/tailwindcss/commands.rb
@@ -74,11 +74,22 @@ def executable(exe_path: DEFAULT_DIR)
       end
 
       def compile_command(debug: false, **kwargs)
+        tailwind_config_template = ERB.new(File.read(Rails.root.join("config/tailwind.config.js")))
+        tailwind_config = Tempfile.new("tailwind-config").tap do |f|
+          f.write(tailwind_config_template.result)
+          if debug
+            f.rewind
+            puts "config:"
+            puts f.read
+          end
+          f.close
+        end
+
         command = [
           executable(**kwargs),
           "-i", Rails.root.join("app/assets/stylesheets/application.tailwind.css").to_s,
           "-o", Rails.root.join("app/assets/builds/tailwind.css").to_s,
-          "-c", Rails.root.join("config/tailwind.config.js").to_s,
+          "-c", tailwind_config.path,
         ]
 
         command << "--minify" unless (debug || rails_css_compressor?)
diff --git a/lib/tailwindcss/engine.rb b/lib/tailwindcss/engine.rb
index 4b9b9fdc..b6049307 100644
--- a/lib/tailwindcss/engine.rb
+++ b/lib/tailwindcss/engine.rb
@@ -2,8 +2,22 @@
 
 module Tailwindcss
   class Engine < ::Rails::Engine
+    class << self
+      def globs
+        Rails.application.config.assets.tailwind_paths.flat_map do |path|
+          [
+            "#{path}/public/*.html",
+            "#{path}/app/helpers/**/*.rb",
+            "#{path}/app/javascript/**/*.js",
+            "#{path}/app/views/**/*.{erb,haml,html,slim}",
+          ]
+        end
+      end
+    end
+
     initializer "tailwindcss.assets" do
       Rails.application.config.assets.precompile += %w( inter-font.css )
+      Rails.application.config.assets.tailwind_paths = [Rails.root]
     end
 
     initializer "tailwindcss.disable_generator_stylesheets" do
```

The changes to an existing engine to work with this would then look like:

```patch
diff --git a/lib/blorgh/engine.rb b/lib/blorgh/engine.rb
index cdf73863..86d0d651 100644
--- a/lib/blorgh/engine.rb
+++ b/lib/blorgh/engine.rb
@@ -1,5 +1,9 @@
 module Blorgh
   class Engine < ::Rails::Engine
     isolate_namespace Blorgh
+
+    config.after_initialize do
+      Rails.application.config.assets.tailwind_paths << Blorgh::Engine.root
+    end
   end
 end
```

If this approach still seems reasonable, then we can talk about what the upgrade experience looks like for existing apps, because the tailwind config would have to change like:

```patch
diff --git a/config/tailwind.config.js b/config/tailwind.config.js
index d6ad82c0..16bbedda 100644
--- a/config/tailwind.config.js
+++ b/config/tailwind.config.js
@@ -1,12 +1,7 @@
 const defaultTheme = require('tailwindcss/defaultTheme')
 
 module.exports = {
-  content: [
-    './public/*.html',
-    './app/helpers/**/*.rb',
-    './app/javascript/**/*.js',
-    './app/views/**/*.{erb,haml,html,slim}'
-  ],
+  content: <%= Tailwindcss::Engine.globs.to_json %>,
   theme: {
     extend: {
       fontFamily: {
```

**Reply by nickpoorman** [reply]
This is great! I also got something working that I've been using for the last few weeks, but I am on my phone so it might take me a day or two to upload it. I think my approach was a little different, as it didn't require any changes to tailwind config but does require running a watcher for each engine to pick up changes.

**Reply by flavorjones** [reply]
@javierjulio Do you think an approach like this could work for ActiveAdmin users? Let's talk about where the friction would be.

**Reply by ianterrell** [reply]
I think this is similar to the approach I took outlined in #108. I like it, and the tempfile approach! I'll adopt a few things from this into our current work.

It would also be nice to be able to optionally customize the specific paths in `globs`; we use ViewComponent and have lots of classes in `app/components` (and there are likely outlier cases where some Tailwind classes are referenced elsewhere). Perhaps something like

```ruby
config.after_initialize do
  Rails.application.config.assets.tailwind_roots << Blorgh::Engine.root
  # and / or:
  Rails.application.config.assets.tailwind_paths << "#{Blorgh::Engine.root}/app/components/**/*"
end
```

(I'm brainstorming names there, the prior `tailwind_paths` turned into `tailwind_roots` above, while `tailwind_paths` above could also be `tailwind_extra_globs`.)

**Reply by javierjulio** [reply]
@flavorjones thank you! I've set up your changes in a branch on a fork. The main issue is as @ianterrell mentioned that additional paths need to be set and/or existing ones need to be changed (e.g. different file extensions). Did you have something in mind or a preference for how to support that? I can give that a try. Perhaps keep the default paths list (globs) as is but let engines specify their own paths that get appended, rather than have to use the engine root into the existing default paths?

**Reply by juan-apa** [reply]
hey @javierjulio I used your fork on an app to compile an engine's classes in the host app. However, I see that it works as intended when the views/classes are inside the `app/` or `public/` directory. However, in my case I need to check the `lib/` directory for styles, I modified the default contents paths to include `"#{path}/lib/**/*.` and that works well.
This use case, raises a different issue, I think it would be nice to have something like what @ianterrell suggests, or having to be more explicit when defining the `tailwind_paths`. e.g.
```ruby
config.after_initialize do
  engine_root = Blorgh::Engine.root
  
  Rails.application.config.assets.tailwind_paths.concat([
    "#{engine_root}/lib/**/*.rb",
    "#{engine_root}/public/*.html",
    "#{engine_root}/app/helpers/**/*.rb",
    "#{engine_root}/app/javascript/**/*.js",
    "#{engine_root}/app/views/**/*.{erb,haml,html,slim}"
  ])
end
```
this approach introduces more verbosity, but can make for a speedier tailwind compile, because it won't have to look through files which the developer knows that tailwind classes won't be used.

I'm happy to help on getting a solution ready.

**Reply by bopm** [reply]
@javierjulio @flavorjones I have some suggestions to approach that brewing here, and I implemented those in the format of [PR on PR](https://github.com/javierjulio/tailwindcss-rails/pull/1).

TL;DR I would prefer a bit less magic, like `.js` files that no longer contain valid JS, or the ability to configure defaults for Organization level specific conventions (like templates or javascript locations), as well as ability to pass custom paths to processing without switching for manually configuring all the paths.

Looking forward for suggestions/feedback.

**Reply by bopm** [reply]
After thinking about that approach overnight, I got a feeling that a better one might be something like [this](https://github.com/bopm/tailwindcss-rails/pull/1), in which we parse original globs configuration and add it to our generated globs. This is closer to convention over configuration as it feels and will not break most of configuration in upgrade.

But it opens another can of worms for cases like [transforming sources](https://tailwindcss.com/docs/content-configuration#transforming-source-files).

**Reply by flavorjones** [reply]
We'll need a different approach when Tailwind v4 lands, because there's no configuration file and it tries to intelligently scan the project directory structure. Not sure how easy that will be to extend to engines located in gems, will need to explore a bit when I have time.

**Reply by bopm** [reply]
> there's no configuration file and it tries to intelligently scan

I find this highly unlikely and recommend checking with @adamwathan for details. Because things like [raw content](https://tailwindcss.com/docs/content-configuration#configuring-raw-content) are being used in custom use-cases like PDF generation and cannot be replaced by just scanning the whole project.

**Reply by flavorjones** [reply]
I appreciate your incredulity, but from https://tailwindcss.com/blog/tailwindcss-v4-alpha:

> You’ll notice that at least in these early alpha releases, it’s not even possible to configure your content paths. For most projects, you’re never going to need to do this ever again — Tailwind just finds your template files for you.

It also says

> We’ll introduce a way to configure content paths explicitly in the future for sure, but we’re curious to see how well this automatic approach works

So maybe if I wait long enough the solution will appear. :wink:

**Reply by bopm** [reply]
Don't get me wrong, but it's not the same as config going away. Ability to configure content paths probably are. But then again, ain't it incredibly wasteful to rescan every gem template on every rebuild? 

I really like you to take a look at the last reply to this discussion.

**Reply by flavorjones** [reply]
The point I am trying to make is: we can't pick a solution until we understand how v4 is going to allow configuration. I said "We'll need a different approach" and "will need to explore a bit" so please don't mistake what I'm saying for anything more than that.

**Reply by bopm** [reply]
It'll allow it in full if it is more complex than [this](https://github.com/tailwindlabs/tailwindcss/pull/14639/files#diff-91efdc92b1c5d77077810317cf1d2cc55890b9a361761a9545be699cde2e29c1R166-R176).

**Reply by bopm** [reply]
@flavorjones After looking at what v4 gave us, seems like you were right, and here is [result](https://github.com/rails/tailwindcss-rails/pull/484) of working towards that. Tell me what I need to do to get it accepted.

**Reply by flavorjones** [reply]
Just wrote up some thoughts in @bopm's PR.

Maybe relevant to readers of this discussion: that PR addresses the use case of "generate one big tailwind.css file for the app and all its mounted engines". And I do think that's the use case this library is well-suited to solve.

But there's another big use case mentioned in this discussion: "each engine has its own tailwind css file." My current mental model is that this "individual engine css" use case can be addressed by using `tailwindcss-ruby` directly, assuming that the engine would package up its pre-built CSS assets in its gem file. Please let me know if that idea doesn't work for you, and if not why not, and please help me update my mental models.

**Reply by bopm** [reply]
> by using tailwindcss-ruby directly

we only need to have some kind of a flag in engine for that to tell our collecting code to not bring that css into main tailwind css.

**Comment by mattalat** [comment]
We have a case where we manage the general front-end (basic assets, configuration, layout partials, some components) in a gem containing a Rails engine. Our teams are small and make many small, separate applications, so extracting most of the front-end matter lets us develop quickly and keeps updates centralized and more manageable.

It would be nice for us to have a way to set up and import a standard tailwind config from the gem and merge it with customizations for each app. There is almost certainly a way to do this without building it into this gem, but I thought I'd mention it anyway.

The more desired feature would be to add our gem's views and components to the paths monitored by this gem.

**Comment by hmasing** [comment]
Adding my $0.02 worth - I found this thread trying to enable precisely this use case for my platform.

Use case: We have one platform, which runs 3 flavors of front end (**internal**, **client**, and **customer checkout**), and on top of that, there's additional white labeling for each **customer checkout** that we need to support. 

Our _application.tailwind.css_ is becoming a bit of a mess of classes, namespaces, etc, and I would really like to split these out.

I am not in a position to directly contribute code, but please count me as a +1 for needing this functionality! 

Thanks, all!

**Comment by nickpoorman** [comment]
I have another use case, similar to some that have already been mentioned. I have multiple websites hosted inside the same rails server process where each website is an engine mounted in the top level `routes.rb`. Each engine makes use of some common engines, ie. `billing_engine`, `authentication_engine`, etc.. are used by `website1_engine` and `website2_engine`.

In fact, I don't have a top level `app` directory and instead have a `core_engine`; see this presentation for the general idea: https://youtu.be/StDoHXO8H6E?si=8KVB6U5yXwHY0CsB

Ideally, I would have an isolated outputted css built for each individual engine as I don't want to load styles that belong to other websites (engines). It would also be nice to be able to reference other engine's, but that's less important to me.

I should also note that I'm using propshaft and not sprockets.

**Comment by Jasonlgrd** [comment]
Hey 👋  I have read all posts to know how I can have two tailwind configurations and I have finish here 😁 

I want to contribute with the description of my use case. I have two applications :
- Back-office : to manage entites and it built with tailwindcss and Flowbite so I need special configuration for this. I use Flowbite to build faster the BO.
- Font-end : Website seeing by classic users with design system made by designer.

I need to dissociate both configurations because the design system is completely different. For the moment I choose to use tailwind without the gem and use https://tailwindcss.com/blog/tailwindcss-v3-2#multiple-config-files-in-one-project-using-config

I can't contribute of the feature development because I'm a noobie in Rails yet but I will very happy if we can use many configurations of Tailwind 🫶

Thanks of lot !

**Comment by javierav** [comment]
My two cents here. In our case we have the main application with its public and private part (it is not relevant here to optimize the styles by separating them into two different files) and then we generate PDFs using HTML and this is where it may be interesting to be able to separate the styles to avoid sending the converter the styles that are not necessary, but I think a bit like dhh, it is more for optimization than a technical necessity.

Thanks for the effort!

**Comment by bopm** [comment]
While waiting for a response to the thread here, I decided to explore different approaches and implement an installation script that collects tailwind classes from `tailwind.css` of my engine and adds them to `raw` configuration of my host app.
https://github.com/full-stack-biz/turbo_material/blob/develop/lib/generators/turbo_material/install_generator.rb#L18-L45

**Comment by DavidGeismarLtd** [comment]
Hey there ! Same need here : 
Multitenant app with different themes (color, spacing...) handled by the same rails app.
So we need to be able to generate different output file based on different tailwind config file.
I guess we thus need the watch process to be able to handle different files based on those different configs.
Has anyone built a workaround to this yet ?

**Reply by flavorjones** [reply]
Hi! Can you say more about your particular use case? It sounds like perhaps each tenant can dynamically configure the style on their (white-labeled) site, but I'd love to confirm that with you.

I'm curious, is there a reason you couldn't use `tailwindcss-ruby` (maybe run in a background job) to generate a CSS file for each tenant as their styles are updated?

**Comment by code-bunny** [comment]
Did we get any traction on engine support. I am off trying it today and expected it to just work out the box until it didn't. What the current best way to work with engines?

**Reply by flavorjones** [reply]
It's not supported at this time. The conversation above includes proposals for a few different approaches we might take for different use cases.

Specifically, in this discussion we're trying to define what problem (or problems) we specifically want to solve. Could you share what you expected "engine support" to mean? I.e., do you expect a distinct CSS file for each engine, or do you expect all engines to be scanned and represented in a single global CSS file (or something else)?

**Reply by flavorjones** [reply]
> What the current best way to work with engines?

One way to do this is to add your engine paths into your `tailwind.config.js` file, which should then include the engines' styles in the global `application.tailwind.css`.

**Reply by code-bunny** [reply]
I can imagine two options:

* Each engine has a tailwind.config that is automatically inherited by the main applications tailwindcss. Given that Tailwindcss should have no conflicting styles, this would allow a 'magic' `stylesheet_link_tag "main_app.tailwind", "inter-font", "data-turbo-track": "reload"` stylesheet for the engine.
* For engines where the author does not intend nor rely on the host to have tailwind we have a `engine_name.tailwind` generated that is fully self contained. However since we still need to run tailwindcss in the Procfile.dev the writer of said engine would be responsible for including this as part of the engine install.

The second option fits my use case the best where I want all my front end in an isolated engine as the main app is a container for three API engines and then the admin engine(which needs tailwindcss). Having the main app cluttered with a stylesheet folder and tailwindcss config isn't great. This also works for when you are mounting several frontend engines in one container app and want tailwind to build isolated stylesheets for each engine so there is no bleeding.

**Comment by flavorjones** [comment]
Hey all, @bopm's pull request #554 adding experimental support for Rails engines has been merged and released in [v4.3.0.rc1](https://github.com/rails/tailwindcss-rails/releases/tag/v4.3.0.rc1). We'd love for you to give it a try and let us know what you think.

**Reply by bopm** [reply]
Thank you @flavorjones for your attention to detail and help along the way.

**Reply by bopm** [reply]
[Here is](https://github.com/full-stack-biz/turbo_material/blob/develop/test/dummy/app/assets/tailwind/application.css#L2) an example of usage in the wild.

**Reply by bopm** [reply]
I found an issue in my approach while deploying with Docker. This [PR](https://github.com/rails/tailwindcss-rails/pull/556) fixes it.

**Reply by tomgallagheradj** [reply]
Hello!

Thanks for all the work on this - it’s a great approach.

I tried integrating it into a Rails engine this week (commit here: https://github.com/adjusted-ai/layered-ui-rails/tree/17188761bb1265f9a7bd5989e150ffaeb1a7bd5989e150ffaeb1a75985) but I’ve hit an issue where the host app doesn’t appear to compile any Tailwind styles for views that originate from the engine.

As a workaround, I’m currently copying the engine’s CSS into the host app via a task (example: https://github.com/adjusted-ai/layered-ui-rails/commit/ed0dcf478bf4bb4d67af9a44a95b756f1ed2b458), but this feels a bit clumsy - I’m assuming there’s a cleaner solution.

I’ve experimented with the @source directive and also considered adding the engine’s view paths to the host app’s Tailwind config. However, that seems like it could lead to brittle `../../../vendor/...` style paths depending on install/setup.

Am I missing part of the recommended setup, or is this something that might be worth discussing as an enhancement? Has anyone successfully used this strategy with engine-provided views?

Happy to contribute if helpful - this would be valuable across a few gems I’m working on.

Many thanks!

**Reply by tibomogul** [reply]
Hello @tomgallagheradj 

I have it working in my project, will share the code soon. There is some discussion here. https://github.com/rails/tailwindcss-rails/issues/565 I think as long as you dont commit the generated css file and treat it as a build artifact it is okay.

**Reply by bopm** [reply]
@tomgallagheradj did you read [this](https://github.com/rails/tailwindcss-rails#rails-engines-support-experimental)? What do you have in your host app in `app/assets/builds/tailwind/layered-ui.css`?

[Here](https://github.com/full-stack-biz/turbo_material/blob/be686de9029ed771fafdd36756185e419fc67da1/test/dummy/app/assets/tailwind/application.css#L2) is how it works in Engine I built it for.

**Reply by thomasgallagher** [reply]
Hello both, apologies for the long delay in getting back to you.

Thanks for the advice. I was able to get this working, and everything appears to be behaving as expected now:
https://github.com/layered-ai-public/layered-ui-rails/commit/3ee706f3aa753beb19d8f4e8d8c7fb9478ac31f2

Thanks again for your help :)
